wallet/psbt.go

```diff
@@ -0,0 +1,298 @@
+//
+// NOTE: If the packet doesn't contain any inputs, coin selection is performed
+// automatically. If the packet does contain any inputs, it is assumed that full
+// coin selection happened externally and no additional inputs are added. If the
+// specified inputs aren't enough to fund the outputs with the given fee rate,
+// an error is returned. No lock lease is acquired for any of the
+// selected/validated inputs. It is in the caller's responsibility to lock the
+// inputs before handing them out.
+func (w *Wallet) FundPsbt(packet *psbt.Packet, account uint32,
+	feeSatPerKB btcutil.Amount) error {
+
+	// Make sure the packet is well formed. We only require there to be at
+	// least one output but not necessarily any inputs.
+	err := psbt.VerifyInputOutputLen(packet, false, true)
+	if err != nil {
+		return err
+	}
+
+	txOut := packet.UnsignedTx.TxOut
+	txIn := packet.UnsignedTx.TxIn
+
+	// Make sure none of the outputs are dust.
+	for _, output := range txOut {
+		// When checking an output for things like dusty-ness, we'll
+		// use the default mempool relay fee rather than the target
+		// effective fee rate to ensure accuracy. Otherwise, we may
+		// mistakenly mark small-ish, but not quite dust output as
+		// dust.
+		err := txrules.CheckOutput(output, txrules.DefaultRelayFeePerKb)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Let's find out the amount to fund first.
+	amt := int64(0)
+	for _, output := range txOut {
+		amt += output.Value
+	}
+
+	// addInputInfo is a helper function that fetches the UTXO information
+	// of an input and attaches it to the PSBT packet.
+	addInputInfo := func() error {
+		packet.Inputs = make([]psbt.PInput, len(txIn))
+		for idx, in := range txIn {
+			utxo, _, err := w.FetchInputInfo(&in.PreviousOutPoint)
+			if err != nil {
+				return fmt.Errorf("error fetching UTXO: %v",
+					err)
+			}
+
+			packet.Inputs[idx].WitnessUtxo = &wire.TxOut{
+				Value:    utxo.Value,
+				PkScript: utxo.PkScript,
+			}
+			packet.Inputs[idx].SighashType = txscript.SigHashAll
+
+			// We don't want to include the witness just yet.
+			packet.UnsignedTx.TxIn[idx].Witness = wire.TxWitness{}
+		}
+
+		return nil
+	}
+
+	var tx *txauthor.AuthoredTx
+	switch {
+	// We need to do coin selection.
+	case len(txIn) == 0:
+		// We ask the underlying wallet to fund a TX for us. This
+		// includes everything we need, specifically fee estimation and
+		// change address creation.
+		tx, err = w.CreateSimpleTx(
+			account, packet.UnsignedTx.TxOut, 1, feeSatPerKB,
+			false,
+		)
+		if err != nil {
+			return fmt.Errorf("error creating funding TX: %v", err)
+		}
+
+		// Copy over the inputs now then collect all UTXO information
+		// that we can and attach them to the PSBT as well. We don't
+		// include the witness as the resulting PSBT isn't expected not
+		// should be signed yet.
+		txIn = tx.Tx.TxIn
+		packet.UnsignedTx.TxIn = tx.Tx.TxIn
+		err = addInputInfo()
```

Comment:
How about just passing the txin as an argument to this func? This way we don't need to mess around with swapping out the variables the closure closes over.